Remove Legacy Metrics

garyrussell opened the pull request:
- Simplify MBeans - instead of wrapping to expose lifecycle methods,
  implement `ManageableLifecycle`. Register an additional MBean for
  polled endpoints to control the lifecycle.

garyrussell committed:
Remove Legacy Metrics

- Simplify MBeans - instead of wrapping to expose lifecycle methods,
  implement `ManageableLifecycle`. Register an additional MBean for
  polled endpoints to control the lifecycle.

artembilan commented on spring-integration-jmx/src/main/java/org/springframework/integration/monitor/IntegrationMBeanExporter.java:
```diff
@@ -348,25 +344,25 @@ private void postProcessAbstractEndpoint(Object bean) {
-				registerHandler(messageHandlerMetrics);
-				this.handlers.add(messageHandlerMetrics);
-				this.runtimeBeans.add(messageHandlerMetrics);
+			if (handler instanceof AbstractMessageHandler) {
```
I think we don't need this `if`. The logic below covers everything we needed.
And having a `MessageHandler` as a proxy won't give us a positive match with this `if` anyway.

garyrussell replied:
Yeah - I went through a lot of iterations with this; I thought I had cleaned all these up. Good catch.

artembilan reviewed (changes requested):
Not too much to say since PR is huge.
Probably I'll merge it as is after your clarifications to my questions.
Mostly I trust you.

Thanks

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/channel/QueueChannel.java:
```diff
@@ -219,11 +219,13 @@ protected boolean doSend(Message<?> message, long timeout) {
 		return purgedMessages;
 	}
 
+	@ManagedAttribute(description = "Queue size")
```
`AbstractMessageChannel` is annotated with `@IntegrationManagedResource`.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/core/Pausable.java:
```diff
@@ -27,16 +28,18 @@
 	/**
 	 * Pause the endpoint.
 	 */
+	@ManagedOperation(description = "Pause the component")
```
We export all components; they don't need `@ManagedResource`; they do, however, need `@IntegrationManagedResource` to prevent the normal MBE from exporting.

That said, pause/resume are not showing up - investigating...

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/support/management/IntegrationManagement.java:
```diff
@@ -84,14 +83,21 @@ default void registerMetricsCaptor(MetricsCaptor captor) {
+	 * @since 5.4
+	 */
+	@SuppressWarnings("unchecked")
+	default <T> T getThisAs() {
```
I understand; this is hard to review; I just thought this would be a convenience to avoid casting.

artembilan replied:
Sure! We can give it a shot!
👍

garyrussell committed:
Polishing

- Move `QueueChannel` `@ManagedAttribute`s to `QueueChannelOperations`
- Make all `AbstractEndpoints` `IntegrationManagedResource`s and remove `ManagedEndpoint`
  to allow exposure of any `@Managed*` methods (including those on `Pausable`)
- Revert to `Lifecycle` for classes that are not related to endpoints
- Remove legacy metrics from docs

garyrussell commented:
Rebased and re-pushed with polishing.
